runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+	type RetirementPeriod: Get<Self::BlockNumber>;
+
+	/// What to do with funds that were not withdrawn.
+	type OrphanedFunds: OnUnbalanced<NegativeImbalanceOf<Self>>;
+}
+
+pub type FundIndex = u32;
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct FundInfo<AccountId, Balance, Hash, BlockNumber, ParaId> {
+	/// The parachain that this fund has funded, if there is one. As long as this is `Some`, then
+	/// the funds may not be withdrawn and the fund cannot be dissolved.
+	parachain: Option<ParaId>,
+	/// The owning account who placed the deposit.
+	owner: AccountId,
+	/// The amount of deposit placed.
+	deposit: Balance,
+	/// The total amount raised.
+	raised: Balance,
+	/// Block number after which the funding must have succeeded. If not successful at this number
+	/// the everyone may withdraw their funds.
+	end: BlockNumber,
+	/// A hard-cap on the amount that may be contributed.
+	cap: Balance,
+	/// The most recent block that this had a contribution. Determines if we make a bid or not.
+	/// If this is `None`, then the last contribution was made outside of the ending period.
+	last_contribution: Option<BlockNumber>,
+	/// First slot in range to bid on; it's actually a LeasePeriod, but that's the same type as
+	/// BlockNumber.
+	first_slot: BlockNumber,
+	/// Last slot in range to bid on; it's actually a LeasePeriod, but that's the same type as
+	/// BlockNumber.
+	last_slot: BlockNumber,
+	/// The deployment data associated with this fund, if any. Once set it may not be reset. First
+	/// is the code hash, second is the initial head data.
+	deploy_data: Option<(Hash, Vec<u8>)>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		/// Info on all of the funds.
+		Funds get(funds):
+			map FundIndex => Option<FundInfo<T::AccountId, BalanceOf<T>, T::Hash, T::BlockNumber, ParaIdOf<T>>>;
+
+		/// The total number of funds that have so far been allocated.
+		FundCount get(fund_count): FundIndex;
+
+		/// The funds that have had additional contributions during the last block. This is used
+		/// in order to determine which funds should submit updated bids.
+		NewRaise: Vec<FundIndex>;
+
+		/// True if the fund was ending at the last block.
+		WasEnding: bool;
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+		where
+	<T as system::Trait>::AccountId,
+	Balance = BalanceOf<T>,
+	ParaId = ParaIdOf<T>,
+	{
+		Created(FundIndex),
+		Contributed(AccountId, FundIndex, Balance),
+		Withdrew(AccountId, FundIndex, Balance),
+		Retiring(FundIndex),
+		Dissolved(FundIndex),
+		DeployDataFixed(FundIndex),
+		Onboarded(FundIndex, ParaId),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+		
+		/// Create a new crowdfunding campaign for a parachain slot deposit.
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
+			#[compact] end: T::BlockNumber,
```

Comment:
Missing check that `end` is greater than `now`.

Should there also be some "minimum" crowdfund time? Would never make sense to have an `end` be one block after `now`.

Reply:
i don't think a prescribing a minimum is necessary: it doesn't provide any additional economic protections and logic to save the users from themselves should be in the UI, not increasing bug surface of the runtime.